File: CS/CustomProgressIndicator/Form1.cs
using System;
using System.Data.OleDb;
using System.IO;
using System.Windows.Forms;
using CustomProgressIndicator.nwindDataSetTableAdapters;
using DevExpress.Snap.Core.API;
using DevExpress.Snap.Core.Services;
using DevExpress.Utils;
// ...

namespace CustomProgressIndicator {
    public partial class Form1 : Form {
        public Form1() {
            InitializeComponent();
        }

        private static object CreateDataSource() {
            var dataSource = new nwindDataSet();
            var connection = new OleDbConnection();
            connection.ConnectionString = Properties.Settings.Default.nwindConnectionString;

            CustomersTableAdapter customers = new CustomersTableAdapter();
            customers.Connection = connection;
            customers.Fill(dataSource.Customers);

            OrdersTableAdapter orders = new OrdersTableAdapter();
            orders.Connection = connection;
            orders.Fill(dataSource.Orders);

            var bindingSource = new BindingSource();
            bindingSource.DataSource = dataSource;
            bindingSource.DataMember = "Customers";
            return bindingSource;
        }

        private void Form1_Load(object sender, EventArgs e) {
            this.snapControl1.DataSources.Add(string.Empty, CreateDataSource());

            this.snapControl1.Options.SnapMailMergeVisualOptions.DataSourceName = string.Empty;
            string filename = FilesHelper.FindingFileName(AppDomain.CurrentDomain.BaseDirectory,
                @"Data\template.snx", true);

            if (File.Exists(filename))
                this.snapControl1.Document.LoadDocument(filename, SnapDocumentFormat.Snap);
            this.snapControl1.ReplaceService<ISnapMailMergeProgressIndicationService>(
                new MyProgressIndicationService(this.snapControl1));
        }
    }
}

File: VB/CustomProgressIndicator/Form1.vb
Imports System
Imports System.Data.OleDb
Imports System.IO
Imports System.Windows.Forms
Imports CustomProgressIndicator.nwindDataSetTableAdapters
Imports DevExpress.Snap.Core.API
Imports DevExpress.Snap.Core.Services
Imports DevExpress.Utils

' ...
Namespace CustomProgressIndicator

    Public Partial Class Form1
        Inherits Form

        Public Sub New()
            InitializeComponent()
        End Sub

        Private Shared Function CreateDataSource() As Object
            Dim dataSource = New nwindDataSet()
            Dim connection = New OleDbConnection()
            connection.ConnectionString = Properties.Settings.Default.nwindConnectionString
            Dim customers As CustomersTableAdapter = New CustomersTableAdapter()
            customers.Connection = connection
            customers.Fill(dataSource.Customers)
            Dim orders As OrdersTableAdapter = New OrdersTableAdapter()
            orders.Connection = connection
            orders.Fill(dataSource.Orders)
            Dim bindingSource = New BindingSource()
            bindingSource.DataSource = dataSource
            bindingSource.DataMember = "Customers"
            Return bindingSource
        End Function

        Private Sub Form1_Load(ByVal sender As Object, ByVal e As EventArgs)
            snapControl1.DataSources.Add(String.Empty, CreateDataSource())
            snapControl1.Options.SnapMailMergeVisualOptions.DataSourceName = String.Empty
            Dim filename As String = FilesHelper.FindingFileName(AppDomain.CurrentDomain.BaseDirectory, "Data\template.snx", True)
            If File.Exists(filename) Then snapControl1.Document.LoadDocument(filename, SnapDocumentFormat.Snap)
            snapControl1.ReplaceService(Of ISnapMailMergeProgressIndicationService)(New MyProgressIndicationService(snapControl1))
        End Sub
    End Class
End Namespace
